server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java

```diff
@@ -38,12 +39,28 @@
 import io.crate.types.ArrayType;
 import io.crate.types.DataType;
 import io.crate.types.TypeSignature;
+import io.crate.types.UndefinedType;
 
 public class ArrayCatFunction extends Scalar<List<Object>, List<Object>> {
 
     public static final String NAME = "array_cat";
 
     public static void register(Functions.Builder module) {
+        // We need to register a concrete function implementation for the `NULL` arguments.
+        // Otherwise, the arguments type will cast to and `ARRAY(UNDEFINED)` which causes a validation error as
+        // the inner type of the array arguments cannot both be `UNDEFINED`.
```

Comment:
Would there be any downsides to allowing `array(undefined)` instead? I could imagine this case occuring more than just once, and it would avoid having the extra registration (plus the confusing error message containing `Possible candidates: array_cat(undefined, undefined):undefined`)

Reply:
> Would there be any downsides to allowing array(undefined) instead?

I see no real issue with allowing both arguments to be `array(undefined)` beside that it will deviate from PG's behaviour which raises and exception in this case. Other than this, it would just work and returns an empty array of type `array(undefined)`

Reply:
Ah I think what I missed was that we want to have the error for `array_cat([], [])` where the inner type is also undefined. 

Maybe rephrase slightly?

```
        // Registration to match `array_cat(null, null)` which should return `null`
        // Without this the signature below (`array_cat(array_undefined, array_undefined)`) would match
        // leading to a validation error which we need to preserve for `array_cat([], [])`
```

Reply:
> Ah I think what I missed was that we want to have the error for array_cat([], []) where the inner type is also undefined.

Do we really want the error?  I see no real need for it. I've changed it to not raise an error, see https://github.com/crate/crate/pull/16295/commits/6357e0b16b238061b76d11cf29b9896786c97de3

Reply:
Also okay I guess. I can't imagine a scenario where somebody would want to rely on that error.